### DISCUSSION TITLE
PropertyDeclarationSyntax and MethodDeclarationSyntax have a common property but don't inherit from the same class or implement the same interface

### DISCUSSION THREAD

**Discussion by awa5114** [dotnet/roslyn#66799] [Q&A]
I'm [crossposting ](https://stackoverflow.com/questions/75401794/propertydeclarationsyntax-and-methoddeclarationsyntax-have-a-common-property-but/75402007?noredirect=1#comment133045353_75402007)from SO
I also [responded ](https://github.com/dotnet/roslyn/issues/9908) to an issue on the main repo but didn't get a satisfactory answer.
If anyone here could provide some ideas or pointers that do not involve using `Reflection` I'm all ears.

In C# syntax analysis, the `PropertyDeclarationSyntax` and `MethodDeclarationSyntax` both have an `Identifier` property. However, they both inherit from `CSharpSyntaxNode` which does not.

I'd like to write a method that can take a `CompilationUnitSyntax` object, extract all `PropertyDeclarationSyntax` or `MethodDeclarationSyntax` objects from it and populate a list with their names. Here's what I have so far:

```
using System;
using System.Collections.Generic;
using System.IO;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System.Linq;

class Program
{

    static List<string> GrabNames(CompilationUnitSyntax root)
    {
        List<MethodDeclarationSyntax> methods = root.DescendantNodes().OfType<MethodDeclarationSyntax>().ToList();
        List<string> methodNames = new List<string>();

        foreach (MethodDeclarationSyntax method in methods)
        {
            methodNames.Add(method.Identifier.Text);
        }

        return methodNames;
    }

    static void Main(string[] args)
    {
        string fileContainingCode = "test.cs"; // file path
        string CodeAsText = File.ReadAllText(fileContainingCode);
        SyntaxTree tree = CSharpSyntaxTree.ParseText(CodeAsText);
        CompilationUnitSyntax root = tree.GetCompilationUnitRoot();

        List<string> names = GrabNames(root);
}
```

I can think of one ugly solution, namely to add a string parameter called `NodeTypeAsString` and based on its value have two branches of an `if` statement do the exact same thing, but then on different types. 

However, I'm thinking there has got to be a better/cleaner way. How can I achieve this without duplicating too much code?

**Comment by CyrusNajmabadi** [comment]
```c#
    static IEnumerable<string> GrabNames<T>(CompilationUnitSyntax root, Func<T, SyntaxToken> getIdentifier) where T : SyntaxNode
        => root.DescendantNodes().OfType<T>().Select(n => getIdentifier(n).ValueText);
```

Then just call it like:

```c#
var methodNames = GrabNames<MethodDeclarationSyntax>(root), m => m.Identifier);
var propertyNames = GrabNames<PropertyDeclarationSyntax>(root), p => p.Identifier);
```

etc.
